aptos-move/framework/aptos-stdlib/sources/cryptography/curves.move

```diff
@@ -0,0 +1,217 @@
+module aptos_std::curves {
+    use aptos_std::type_info::type_of;
+
+    // Structs and consts.
+
+    // Fake structs representing group type.
+    struct BLS12_381_G1 {}
+    struct BLS12_381_G2 {}
+    struct BLS12_381_Gt {}
+
+    //TODO: handle as u8 temporarily. Upgrade to u64.
+    struct Scalar<phantom Group> has drop {
+        handle: u8
+    }
+
+    struct Point<phantom Group> has drop {
+        handle: u8
+    }
+
+    /// Get internal handle for a Scalar. Currently needed by groth16 module.
+    /// TODO: can this be avoided?
+    public fun get_scalar_handle<Group>(s: &Scalar<Group>): u8 {
+        s.handle
+    }
+
+    /// Get internal handle for a point. Currently needed by groth16 module.
+    /// TODO: can this be avoided?
+    public fun get_point_handle<Group>(p: &Point<Group>): u8 {
+        p.handle
+    }
+
+    /// Perform a bilinear mapping.
+    /// TODO: is it possible to have 2+ mappings between same (G1,G2,Gt)? If so we need a parameter for `mapping_id`?
+    public fun pairing<G1,G2,Gt>(point_1: &Point<G1>, point_2: &Point<G2>): Point<Gt> {
+        Point<Gt> {
+            handle: pairing_internal(point_1.handle, point_2.handle, get_pairing_id<G1,G2,Gt>())
+        }
+    }
+
```

Comment:
Also add a optimized `pairing_product` API that uses Miller loops to speed-up computations:

```
public fun pairing_product<G1,G2,Gt>(points_1: &vector<Point<G1>>, points_2: &vector<Point<G2>>): Point<Gt>
```